packages/flutter_tools/lib/src/ios/devices.dart

```diff
@@ -288,12 +291,15 @@ class IOSDevice extends Device {
         printError('');
         return LaunchResult.failed();
       }
+      packageId = buildResult.xcodeBuildExecution?.buildSettings['PRODUCT_BUNDLE_IDENTIFIER'];
     } else {
       if (!await installApp(package)) {
         return LaunchResult.failed();
       }
     }
 
+    packageId ??= package.id;
```

Comment:
Looking at this again, does `iosApp.id` have the value we want instead?  It looks like it's populated by parsing the plist AND falling back to the PRODUCT_BUNDLE_IDENTIFIER. https://github.com/flutter/flutter/blob/08c645b3b2da8c66b945fe701be99c52c0d11f5b/packages/flutter_tools/lib/src/project.dart#L381

Reply:
AND the test I attached passes on master!  Is it possible `package.id` is already pulling from the PRODUCT_BUNDLE_IDENTIFIER?

Reply:
I don' think so.
The iosApp is just the "package", that gets passed into the method from **before** the app is built.
```
    // Step 2: Check that the application exists at the specified path.
    final IOSApp iosApp = package;
```

In this case, `Info.plist` does not exist yet, so it falls back to parsing `project.pbxproj`. 
However, since this file holds the build configuration for all build schemes, it just returns the **first match** of PRODUCT_BUNDLE_IDENTIFIER which is not necessarily the scheme with which the app is going to be built.

It seems that setting the default configuration (see https://imgur.com/a/atu7Kki) "Use <Scheme> for command-line builds" reorders to entries in `project.pbxproj`, which is why it works if you set this to the configuration which gets used by flutter in the following build (this gets mentioned here as well: https://github.com/flutter/flutter/issues/18342#issuecomment-396881069 )

Reply:
I'll try to come up with a test case that fails on master and succeeds with my fixes, but it will take some time for me to understand the testing architecture.

Reply:
The test for this is not going to be trivial.
In order to really test this, we need to build the app with a custom flavor/scheme that overwrites the $PRODUCT_BUNDLE_IDENTIFIER.
And then start the app like this:

```dart
device.startApp(
              app,
              prebuiltApplication: false,
              debuggingOptions: DebuggingOptions.disabled(const BuildInfo(BuildMode.debug, **"testflavor"**)),
              platformArgs: <String, dynamic>{},
            )
```

I can't open the ios.tmpl directory in xcode though. Can you give me some pointers on how this templating system works? I could not find any documentation for this.

Reply:
> returns the first match of PRODUCT_BUNDLE_IDENTIFIER
Yikes, you're right.

The test I sent doesn't even exercise the code you changed, it only actually gets to the -getObservatoryUri codepath when `DebuggingOptions.enabled` and `prebuiltApplication = false`.
I didn't see a great way to test this either, which probably hints that we should refactor startApp to be more testable.
Honestly what you have is fine, I really appreciate you being thoughtful about this change!